Content.Server/Objectives/ObjectivesManager.cs

```diff
@@ -10,21 +11,31 @@ public sealed class ObjectivesManager : IObjectivesManager
         [Dependency] private IPrototypeManager _prototypeManager = default!;
         [Dependency] private IRobustRandom _random = default!;
 
-        public IEnumerable<ObjectivePrototype> GetAllPossibleObjectives(Mind.Mind mind)
+        public ObjectivePrototype? GetRandomObjective(Mind.Mind mind, string objectiveGroupProto)
         {
-            return _prototypeManager.EnumeratePrototypes<ObjectivePrototype>().Where(objectivePrototype => objectivePrototype.CanBeAssigned(mind));
-        }
-
-        public ObjectivePrototype? GetRandomObjective(Mind.Mind mind)
-        {
-            var objectives = GetAllPossibleObjectives(mind).ToList();
-            _random.Shuffle(objectives);
+            if (!_prototypeManager.TryIndex<WeightedRandomPrototype>(objectiveGroupProto, out var groups))
+            {
+                Logger.Error("Tried to get a random objective, but can't index WeightedRandomPrototype " + objectiveGroupProto);
+                return null;
+            }
 
-            //to prevent endless loops
-            foreach (var objective in objectives)
+            // yeah the old 'preventing infinite loops' thing wasn't super elegant either and it mislead people on what exactly it did
+            var tries = 0;
+            while (tries < 20)
```

Comment:
im not really a fan of this but i also dont see a way that this is gonna work otherwise. i guess you could return a 'default objective' like stay alive or something?

Reply:
the issue is that there's no objectives that are always valid, if we just return stay alive when the user already has die, thenw hat?